rust/lancedb/src/embeddings.rs

```diff
@@ -0,0 +1,264 @@
+use lance::arrow::RecordBatchExt;
+use std::{
+    borrow::Cow,
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+
+use arrow_array::{Array, RecordBatch, RecordBatchReader};
+use arrow_schema::{DataType, Field, SchemaBuilder};
+// use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    error::Result,
+    table::{ColumnDefinition, ColumnKind, TableDefinition},
+};
+
+/// Trait for embedding functions
+///
+/// An embedding function is a function that is applied to a column of input data
+/// to produce an "embedding" of that input.  This embedding is then stored in the
+/// database alongside (or instead of) the original input.
+///
+/// An "embedding" is often a lower-dimensional representation of the input data.
+/// For example, sentence-transformers can be used to embed sentences into a 768-dimensional
+/// vector space.  This is useful for tasks like similarity search, where we want to find
+/// similar sentences to a query sentence.
+///
+/// To use an embedding function you must first register it with the `EmbeddingsRegistry`.
+/// Then you can define it on a column in the table schema.  That embedding will then be used
+/// to embed the data in that column.
+// #[async_trait]
+pub trait EmbeddingFunction: std::fmt::Debug + Send + Sync {
+    fn name(&self) -> &str;
+    fn source_type(&self) -> Cow<DataType>;
+    fn dest_type(&self) -> Cow<DataType>;
```

Comment:
I think the `Cow` return types provide a bit more flexibility to end users as they don't need to go through extra steps to create a reference.

For reference, I opened up an [issue in datafusion](https://github.com/apache/datafusion/issues/10018) a while back highlighting the issues with using borrowed types